logger.go

```diff
@@ -297,11 +297,17 @@ func (log *Logger) check(lvl zapcore.Level, msg string) *zapcore.CheckedEntry {
 	// Thread the error output through to the CheckedEntry.
 	ce.ErrorOutput = log.errorOutput
 	if log.addCaller {
-		ce.Entry.Caller = zapcore.NewEntryCaller(runtime.Caller(log.callerSkip + callerSkipOffset))
-		if !ce.Entry.Caller.Defined {
+		pc := make([]uintptr, 1)
+		numFrames := runtime.Callers(log.callerSkip+callerSkipOffset+1, pc)
+		frame, _ := runtime.CallersFrames(pc[:numFrames]).Next()
+		defined := frame.PC != 0
+		if !defined {
```

Comment:
couple of things here:
 - do we need to handle the case where `numFrames == 0` which is handled [in the runtime](https://golang.org/src/runtime/extern.go?s=9752:9818#L188) (e.g., if someone sets skip too high)
 - let's move the logic into a separate function that returns a frame

Can we also add a comment for the `frame.PC != 0` case (i think that's the "failed to decode case" according to [the docs](https://golang.org/src/runtime/extern.go?s=9752:9818#L183))

Reply:
I completely missed that case... thanks for the suggestions, updated the code @prashantv 
I've checked the benchmark, there should be no additional performance impact